docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,142 @@
+- via object files. This happens when the allocator is not a separate static library.
+- via a separate allocator static library. In this case, the allocator library must be specified before `oelibc` in the linker command line. </br>
+  ELF  `weak symbols` (see [Function Attributes](https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Function-Attributes.html)) are designed to allow user code to override functions defined in libraries.
+  Using `weak symbols` for default implementation of these function would cause the linker to look for custom replacement functions irrespective of the order.
+
+
+Replacement functions are not required to be reentrant.
+
+Undefined Behavior
+---
+
+- The replacement functions should not do much other than managing memory. Making ocalls, logging, printf etc may use dynamic memory and hence result in reentranct calls to the allocator.
+- All the replacement functions must be provided, or none. A partial set will result in both the default allocator and the custom allocator managing the heap.
+
+
+Replacement Functions
+---
+
+Published in `openenclave/enclave/allocator.h`
+
+- `void oe_allocator_init(void* heap_start_address, void* heap_end_address)` </br>
+  This function will be called by `oecore` at an appropriate time during enclave initialization. The total heap memory available to the enclave is specified in its configuration file
+  (e.g [hello_worlf.conf](https://github.com/openenclave/openenclave/blob/e2e440ec134d37e107b5a296fdc508a3a643598e/samples/helloworld/enclave/helloworld.conf#L6)) </br>
+  `oecore` may reserve a small portion for its own use and pass the rest to this replacement function. The allocator is expected to perform any global initialization in this function. </br>
+  This function is called prior to any global initialization.
+
+
+- `void oe_allocator_cleanup(void)` </br>
+  This function will be called by `oecore` at an appropriate time during enclave termination. `atexit` functions will be executed prior to `oe_allocator_cleanup`. </br>
+  `oe_allocator_cleanup` may not register functions with `atexit`.
+
+
+- `void oe_allocator_thread_init(void)` </br>
+  This function will be called by `oecore` during initialization of an enclave thread, after the thread-local variables for the thread have been initialized. </br>
+  In the future, if `oecore` needs to hold onto any thread-specific data on behalf of the allocator, it can be returned in this function. Currently the recommendation is to use `__thread` variables to store thread-specific allocator data.
+
+
+- `void oe_allocator_thread_cleanup(void)` </br>
+  This function will be called by `oecore` just prior to the termination of enclave thread. Thread-specific exit functions (`__cxa_thread_atexit` functions) will be executed prior to `oe_allocator_thread_cleanup`. </br>
+  `oe_allocator_thread_cleanup` may not use `__cxa_thread_atexit`; that is it cannot access C++ thread-local variables with non-trivial destructors for the first time. </br>
```

Comment:
After multiple attempts at wording which required explaining C++11's intricate details, here is what I came up with:

- `void oe_allocator_thread_cleanup(void)` </br>
This function will be called by `oecore` just prior to the termination of enclave thread. 
  Thread-specific exit functions (`__cxa_thread_atexit` functions) will be executed prior to `oe_allocator_thread_cleanup`. </br>
  This allows the thread-specific exit functions to use memory management functions (e.g to free resources) before the allocator itself is cleaned up. </br>
  It is recommended that allocators be judicious about using C++11 thread-specific objects in `oe_allocator_thread_cleanup` since 
  use of C++11 thread-local objects may trigger a call to (`__cxa_thread_atexit`) after all the thread-specific exit functions have executed.